File: js/Chess/Board.js
import Location from './Location'
import TemporaryPieceMove from './TemporaryPieceMove'

export default class Board {
    constructor(cells) {
        this.cells = cells;
    }

    getPieces(color) {
        return this.allBoard().filter(loc => this.getPiece(loc) && this.getPiece(loc).color === color);
    }

    getPiece(location) {
        return this.contains(location) ? this.cells[location.x][location.y] : null;
    }

    set(location, cell) {
        this.cells[location.x][location.y] = cell;
    }

    performTemporaryMove(from, to) {
        const old = this.getPiece(to);
        this.set(to, this.getPiece(from));
        this.set(from, null);
        return new TemporaryPieceMove(this, from, to, old);
    }

    allBoard() {
        const result = [];

        for (let y = 0; y < this.cells.length; y++) {
            for (let x = 0; x < this.cells[0].length; x++) {
                result.push(new Location(x, y));
            }
        }

        return result;
    }

    contains(location) {
        return location.x >= 0 && location.x < this.cells[0].length &&
            location.y >= 0 && location.y < this.cells.length;
    }
}

File: cs/Chess/Board.cs
using System.Collections.Generic;
using System.Linq;

namespace Chess
{
    public class Board
    {
        private readonly Piece[][] cells;

        public Board(Piece[][] cells)
        {
            this.cells = cells;
        }

        public IEnumerable<Location> GetPieces(PieceColor color) => 
            AllBoard().Where(loc => Piece.Is(GetPiece(loc), color));

        public Piece GetPiece(Location location) => 
            Contains(location) ? cells[location.Y][location.X] : null;

        public void SetPieceToLocation(Location location, Piece cell) => 
            cells[location.Y][location.X] = cell;

        public TemporaryPieceMove PerformTemporaryMove(Location departure, Location destination)
        {
            var oldPiece = GetPiece(destination);
            SetPieceToLocation(destination, GetPiece(departure));
            SetPieceToLocation(departure, null);
            
            return new TemporaryPieceMove(this, departure, destination, oldPiece);
        }

        private IEnumerable<Location> AllBoard()
        {
            for (var y = 0; y < cells.Length; y++)
            for (var x = 0; x < cells[0].Length; x++)
                yield return new Location(x, y);
        }

        public bool Contains(Location loc) =>
            loc.X >= 0 
            && loc.X < cells[0].Length 
            && loc.Y >= 0 
            && loc.Y < cells.Length;
    }
}